File: ttcp.cc
#include <iostream>
#include <string>
#include <chrono>
#include "common.h"
#include "message.h"
#include "accept.h"
#include "inet_address.h"

using std::cout;
using std::endl;
using std::string;

struct commandLine
{
   StringArg host;
   StringArg port;
   int bsNumber;
   int bsLength; 
   bool isClinet;
   commandLine() :bsNumber(0), bsLength(0), isClinet(false) {}
   ~commandLine() = default;
};

commandLine* parseCommandLine(int& argc, const char* argv[]) 
{


   if (argc < 3) {
      cout << "Usage -t client/server" << endl;
      return nullptr;
   }

   if (strcmp(argv[1], "-t") != 0) {
      cout << "Usage -t [client/server]" << endl;
      return nullptr;
   }

   if (strcmp(argv[2], "client") == 0) {
      if (argc < 6) {
         cout << "Usage -t client -s host -p port (default -n 1000 -b 1024)" << endl;
         return nullptr;
      }

      
      StringArg host, port;
      int bsCount, bsLength;
      for (int i = 3; i < argc; i++) {
         if (strcmp(argv[i], "-s") == 0 && i+ 1 < argc) {
            host = argv[++i];
         } else if (strcmp(argv[i], "-p") == 0 && i + 1 < argc) {
            port = argv[++i];
         } else if (strcmp(argv[i], "-n") == 0 && i+ 1 < argc) {
            bsCount = atoi(argv[++i]);
         } else if (strcmp(argv[i], "-b") == 0 && i+ 1 < argc) {
            bsLength = atoi(argv[++i]);
         }
      }

      if (port.empty()) {
         cout << "-p port argument empty" << endl;
         return nullptr;
      }

      if (host.empty()) {
         host = "0.0.0.0"; // INADDR_ANY
      }

      if (bsCount == 0) {
         bsCount = 1000;
      }

      if (bsLength == 0) {
         bsLength = 1024;
      }

      struct commandLine* line = new commandLine;
      line->host = host;
      line->port = port;
      line->bsNumber = bsCount;
      line->bsLength = bsLength;
      line->isClinet = true;

      cout << "host: " << line->host.get_c_str()
         << ", port: " << line->port.get_c_str()
         << ", bs count: " << line->bsNumber 
         << ", bs length: " << line->bsLength
         << ", is client: " << line->isClinet << endl;

      return line;
   }
   
   if (strcmp(argv[2], "server") == 0) {
      if (argc < 6) {
         cout << "Usage -t client -s host -p port (default -n 1000 -b 1024)" << endl;
         return nullptr;
      }
      StringArg host, port;
      for (int i = 3; i < argc; i++) {
         if (strcmp(argv[i], "-s") == 0 && i + 1 < argc) {
            host = argv[i+1];
         } else if (strcmp(argv[i], "-p") == 0 && i + 1 < argc) {
            port = argv[i+1];
         }
      }

      if (port.empty()) {
         cout << "-p port argument empty" << endl;
         return nullptr;
      }

      if (host.empty()) {
         host = "0.0.0.0"; // INADDR_ANY
      }

      struct commandLine* line = new commandLine;
      line->host = host;
      line->port = port;
      line->isClinet = false;
      cout << "host: " << line->host.get_c_str()
         << ", port: " << line->port .get_c_str()
         << ", is client: " << line->isClinet << endl;
      return line;
   }

   cout << "invalid -t argument" << endl;

   return nullptr;
}

void transmit(struct commandLine* linePtr) 
{
   InetAddress serverAddr(linePtr->host, static_cast<in_port_t>(linePtr->port.toi()));
   std::unique_ptr<TcpStream> stream = TcpStream::connect(serverAddr);
   
   struct SessionMessage sessionMessage {0,0};
   sessionMessage.number = htonl(linePtr->bsNumber);
   sessionMessage.length = htonl(linePtr->bsLength);
   stream->sendAll(&sessionMessage, sizeof sessionMessage);
   
   int payloadMessageLen = sizeof(uint32_t) + linePtr->bsLength;
   PayloadMessage* payloadMessage = static_cast<PayloadMessage*>(::malloc(payloadMessageLen)); 
   assert(payloadMessage);

   payloadMessage->length = htonl(linePtr->bsLength);
   for (int i = 0; i < linePtr->bsLength; i++) {
      payloadMessage->data[i] = "0123456789ABCDEF"[i % 16];
   }

   double totalMib = 1.0 * linePtr->bsNumber * linePtr->bsLength / 1024 / 1024;
   std::printf("%.3f MiB in total\n", totalMib);

   int nr, wr;
   uint32_t ack = 0;
   for (int i = 0; i < linePtr->bsNumber; i++) {
      wr = stream->sendAll(payloadMessage, payloadMessageLen);
      if (wr != payloadMessageLen) {
         perror("sendAll error");
         abort();
      }

      nr = stream->receiveAll(&ack, sizeof ack);
      if (nr != sizeof ack) {
         perror("receiveAll error");
         abort();
      }

      ack = ntohl(ack);
      if (ack != linePtr->bsLength) {
         perror("server receiveAll error, ack not euqal");
         abort();
      }
   }
}

void receive(struct commandLine* linePtr) 
{
   Acceptor acceptor(InetAddress(linePtr->host, 
      static_cast<in_port_t>(linePtr->port.toi())));
   
   std::unique_ptr<TcpStream> stream = acceptor.accept();

   struct SessionMessage sessionMessage{0, 0};
   int read_n = stream->receiveAll(&sessionMessage, sizeof sessionMessage);
   sessionMessage.number = ntohl(sessionMessage.number);
   sessionMessage.length = ntohl(sessionMessage.length);
   std::printf("receive buffer length = %d\nreceive number of buffers = %d\n",
         sessionMessage.length, sessionMessage.number);

   int payloadLen = sizeof(uint32_t) + sessionMessage.length;
   PayloadMessage *payloadMessage = static_cast<PayloadMessage*>(::malloc(payloadLen));

   double totalMib = 1.0 * sessionMessage.number  * sessionMessage.length / 1024 / 1024;
   std::printf("%.3f MiB in total\n", totalMib);
   
   int nr, wr;
   uint32_t ack;
   auto before = std::chrono::system_clock::now();
   for (int i = 0; i < sessionMessage.number; i++) {
      nr = stream->receiveAll(&payloadMessage->length, sizeof payloadMessage->length);
      if (nr != sizeof payloadMessage->length) {
         cout << "receive error" << endl;
      }

      payloadMessage->length = ntohl(payloadMessage->length);
      if (payloadMessage->length != sessionMessage.length) {
         cout << "receive PayloadMessage length error" << endl;
         return;         
      }

      nr = stream->receiveAll(payloadMessage->data,  payloadMessage->length);
      if (nr != payloadMessage->length) {
         cout << "receive error" << endl;
         return;         
      }

      ack = htonl(payloadMessage->length);
      wr = stream->sendAll(&ack, sizeof ack);
      if (wr != sizeof ack) {
          cout << "send ack error" << endl;
         return;  
      }
   }

   std::chrono::duration<double> elapsed = std::chrono::system_clock::now() - before;
   std::printf("%.3f seconds\n%.3f MiB/s\n", elapsed, totalMib / elapsed.count());
}
// -t client -s host -p port -n 10 -b 1000
// -t server -s host -p port
int main(int argc, const char* argv[]) {
   
   struct commandLine* linePtr = parseCommandLine(argc, argv);
   if (linePtr == nullptr) {
      return 0;
   }
   
   if (linePtr->isClinet) {
      transmit(linePtr);
   }
   
   if (!linePtr->isClinet) {
      receive(linePtr);
   }
   
   delete linePtr;
}

File: python/ttcp.py
import socket
import sys
import struct
import time

class AckMessage():
    def size(self):
        return 4
    
    def pack(self, value):
        return struct.pack('=I', socket.htonl(value))
    
    def unpack(self, data):
        return socket.ntohl(struct.unpack('=I', data)[0])
    
class SessionMessage():
    def __init__(self, number = 0, length = 0):
        self.number = number
        self.length = length
    
    def pack(self):
        return struct.pack('=2I', socket.htonl(self.number), socket.htonl(self.length))

    def unpack(self, data):
        self.number = socket.ntohl(struct.unpack('=I', data[:4])[0])
        self.length = socket.ntohl(struct.unpack('=I', data[4:8])[0])

    def message_length(self):
        return 8

class PayloadMessage():
    def __init__(self, length = 0):
        self.data = b''
        for i in range (length):
            self.data += "0123456789ABCDEF"[i%16].encode()
        self.length = length
    
    def pack(self):
        return struct.pack('=I%ds' % self.length, socket.htonl(self.length), self.data) 

    def unpack_length(self, data):
        self.length = socket.ntohl(struct.unpack('=I', data[:4])[0])

    def length_size(self):
        return 4

    def message_length(self):
        return 4 + self.length

    def unpack_ack(self, ack_data):
        return socket.ntohl(struct.unpack('=I', ack_data)[0])

class TcpStream():
    def __init__(self, so):
        self.so = so
    
    def send_all(self, buf, n):
        sendn = 0
        nw = 0
        while sendn < n:
            nw = self.so.send(buf[sendn:n])
            if nw == 0:
                return 0
            
            if nw < 0:
                return -1

            if nw > 0:
                sendn += nw
        
        return sendn

    def recv_all(self, n):
        return self.so.recv(n, socket.MSG_WAITALL)

    def close(self):
        self.so.close()

def transmit(server_socket, number, length):
    # send session message
    sessionMsg = SessionMessage(number, length)
    server_socket.send(sessionMsg.pack())
    
    # paylaod message
    ackMsg = AckMessage()
    stream = TcpStream(server_socket)
    payloadMsg = PayloadMessage(length)
    
    for _ in range(number):
        # send paylaod message to server
        nw = stream.send_all(payloadMsg.pack(), payloadMsg.message_length())
        if nw < 0:
            print ("stream.send_all payload message fail")
            stream.close()
            return

        if nw == 0:
            print ("stream.send_all payload message fail, server closed.")
            stream.close()
            return

        if nw != payloadMsg.message_length():
            print("stream.send_all payload message fail, actually need send: %d, but strem send: %d" % (payloadMsg.message_length(), nw))
            stream.close()
            return
       
        # recv ack message from server
        data = stream.recv_all(ackMsg.size())
        if len(data) == 0:
            print("stream.recv_all ack message fail,server closed")
            stream.close()
            return
        ack = ackMsg.unpack(data)
        if ack != length:
            print("check ack fail, actually need send length: %d, but server ack recv length: %d" % (length, ack))
            stream.close()
            return

    stream.close()

def receive(server_socket, client_socket):
    # recv session message and unpack from transmit
    sessionMsg = SessionMessage()
    stream = TcpStream(client_socket)
    data = stream.recv_all(sessionMsg.message_length())
    if len(data) != sessionMsg.message_length():
        stream.close()
        print("stream.recv_all error, need recv: %d, give: %d" % sessionMsg.message_length(), len(data))
        return
    
    sessionMsg.unpack(data)

    print("receive buffer length = %d\nreceive number of buffers = %d" % (sessionMsg.length, sessionMsg.number))
    totalMib = sessionMsg.number * sessionMsg.length * 1.0 / 1024 / 1024
    print("%.3f MiB in total" % (totalMib))

    ackMsg = AckMessage()

    # recv payload message and unpack from transmit
    payloadMsg = PayloadMessage()
    start_time = time.time()
    for _ in range (sessionMsg.number):
        # recv payload message length data and unpack it
        data = stream.recv_all(payloadMsg.length_size())
        nr = len(data)
        if nr == 0:
            print("client closed.")
            stream.close()
            return
        
        if nr < 0:
            print("stream.recv_all payload message length data fail.")
            stream.close()
            return
        
        if nr != payloadMsg.length_size():
            print("stream.recv_all payload message length data fail.")
            stream.close()
            return
        
        payloadMsg.unpack_length(data)

        # recv payload message data data and unpack it
        data = stream.recv_all(payloadMsg.length)
        nr = len(data)
        if nr == 0:
            print("client closed.")
            stream.close()
            return

        if nr < 0:
            print("stream.recv_all payload message data data fail.")
            stream.close()
            return

        if nr != payloadMsg.length:
            print("stream.recv_all payload message data data fail, need %d length data, give %d" % payloadMsg.length, nr)

        # send ack of recive data length to client
        nw = stream.send_all(ackMsg.pack(payloadMsg.length), ackMsg.size())
        if nw == 0:
            print("stream.send_all ack message fail, client closed")
            stream.close()
            return
        
        if nw < 0:
            print("stream.send_all ack message fail")
            stream.close()
            return

        if nw != ackMsg.size():
            print("stream.send_all ack message fail, need %d send, give %d" % nw, ackMsg.size())
            stream.close()
            return

    end_time = time.time()
    elapsed = end_time - start_time
    print("%.3f seconds\n%.3f MiB/s\n" % (elapsed, totalMib / elapsed))
    stream.close()

def main(argv):
    if len(argv) < 3:
        print ("Usage:\n -t (client or server)")
        return

    if argv[1] != "-t":
        print ("Usage:\n -t (client or server)")
        return

    if argv[2] == "server":
        host = ""
        port = 0
        for i in range(len(argv)):
            if argv[i] == "-h" and i + 1 < len(argv):
                host = argv[i+1]
            if argv[i] == "-p" and i + 1 < len(argv):
                port = int(argv[i+1])

        if host == "":
            print ("Usage:\n -t server -h host")
            return

        if port == 0:
            print ("Usage:\n -t server -h %s -p port" % (host))
            return

        print ("ttcp listen server %s:%d" % (host, port))
        
        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server_socket.setsockopt(socket.SOL_SOCKET, socket.TCP_NODELAY, 1)
        server_socket.bind((host, port))
        server_socket.listen(5)
        (client_socket, client_address) = server_socket.accept()
        receive(server_socket, client_socket)
        server_socket.close()
  
    if argv[2] == "client":
        host = ""
        port = 0
        number = 1000
        length = 1024
        for i in range(len(argv)):
            if argv[i] == "-h" and i + 1 < len(argv):
                host = argv[i+1]
            if argv[i] == "-p" and i + 1 < len(argv):
                port = int(argv[i+1])
            if argv[i] == "-n" and i + 1 < len(argv):
                number = int(argv[i+1])
            if argv[i] == "-b" and i + 1 < len(argv):
                length = int(argv[i+1])

        if host == "":
            print ("Usage:\n -t client -h host")
            return

        if port == 0:
            print ("Usage:\n -t client -h %s -p port" % (host))
            return
        
        print ("ttcp connection server %s:%d:\n  message number: %d count\n  message length: %d byte\n  total: %.3f Mib" % (host, port, number, length, 
            float(1.0 * number * length / 1024 / 1024)) )

        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server_socket.setsockopt(socket.SOL_SOCKET, socket.TCP_NODELAY, 1)
        server_socket.connect((host, port))
        transmit(server_socket, number, length)

if __name__ == "__main__":
    main(sys.argv)